docs: make a documentation page outlining the terminology based on the spreadsheet (#572)
AutoResearch/autora: docs/terminology.md

# AutoRA Terminology

The following table includes naming conventions used throughout AutoRA.

| Term               | Description                                                                                                                                 | Relevant Modules                                |
|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------|

> | State              | Base object that has all relevant fields and which uses the Delta mechanism to modify those fields.                                         | Core                                            |

Reviewer: Is there a need to mention the Delta mechanism? It doesn't come up in documentation or tutorials, but that's maybe because we always use the StardardState as far as I've seen.

Author: good point. i was going off this: https://docs.google.com/presentation/d/1OdP4M2ByW8GKTFvcCgDG5HDzUwHKdFEhj1ddH4ClvEw/edit#slide=id.g27e19158dea_1_0

which is what we agreed to do in the task: https://github.com/AutoResearch/autora/issues/516#issuecomment-1716300791

Reviewer: I think maybe you misunderstood me. I think having the State here is a good idea, but in the description you say "which uses the Delta mechanism" but, unless I am mistaken, we don't talk about the Delta mechanism anywhere. If we want to include it here, we might also want to add it as a description in this table. I am happy with whatever here though.

| StandardState      | An optional default State that has the following fields: variables, conditions, experiment_data, models.                                    | Core                                            |
| Variables          | A State field that holds experimental variables, which are defined according to name, type, units, allowed values, and range.               | Experimentalists, Experiment Runners, Theorists |
| VariableCollection | Immutable metadata about dependent variables, independent variables, and covariates.                                                        | Experimentalists, Experiment Runners, Theorists |
| Conditions         | A State field that defines what observations should be collected according to a specific combination of values of the independent variables | Experimentalists, Experiment Runners, Theorists |

> | Experiment Data    | A State field that holds observations which correspond to the specified conditions.                                                         | Experiment Runners, Theorists                   |

Reviewer: The Experiment Data field also holds the conditions.

> | Experiment Data    | A State field that holds observations which correspond to the specified conditions.                                                         | Experiment Runners, Theorists                   |

Author: ```suggestion
| Experiment Data    | A State field that holds specified conditions as well as the corresponding observations.                                                         | Experiment Runners, Theorists                   |
```

| Model              | A State field that holds the the collection of best fit equations produced by theorists.                                                    | Theorists, Experimentalists                     |

> | Components         | The distinct yet flexible capabilities of the AutoRA framework.                                                                             | Experimentalists, Experiment Runners, Theorists |

Reviewer: This isn't very descriptive. Might want to change to something like "the objects that can act on the state - e.g., experimentalists, experiment runners, theorists."?

> | Components         | The distinct yet flexible capabilities of the AutoRA framework.                                                                             | Experimentalists, Experiment Runners, Theorists |

Author: ```suggestion
| Components         | The objects that can act on the State (e.g., experimentalists, experiment runners, theorists).                                                                             | Experimentalists, Experiment Runners, Theorists |
```

> | Experimentalist    | A module that takes in models and outputs new conditions, which are intended to yield novel observations.                                   | Experimentalists                                |

Reviewer: I like the word module here and in the next two, but with Components explicitly defined, should we use that terminology instead?

Reviewer: Also, not all experimentalists take in models.

Author: we should maybe update the core diagram that we use throughout the user guide then: https://autoresearch.github.io/autora/img/experimentalist.png

> | Experimentalist    | A module that takes in models and outputs new conditions, which are intended to yield novel observations.                                   | Experimentalists                                |

Author: ```suggestion
| Experimentalist    | A component that outputs new conditions, which are intended to yield novel observations.                                   | Experimentalists                                |
```

> | Theorist           | A module that takes in the full collection of conditions and observations and outputs equations that link the two (i.e., models)            | Theorists                                       |

Reviewer: I would say outputs models instead of equations because they don't always have to output an equation (although all of ours currently do).

> | Theorist           | A module that takes in the full collection of conditions and observations and outputs equations that link the two (i.e., models)            | Theorists                                       |

Author: ```suggestion
| Theorist           | A component that takes in the full collection of conditions and observations and outputs models that link the two.            | Theorists                                       |
```

> | Experiment Runner  | A module that takes in conditions and collects corresponding observations.                                                                  | Experiment Runners                              |

Author: ```suggestion
| Experiment Runner  | A component that takes in conditions and collects corresponding observations.                                                                  | Experiment Runners                              |
```

# AutoRA Terminology

The following table includes naming conventions used throughout AutoRA.

| Term               | Description                                                                                                                                 | Relevant Modules                                |
|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------|

> | State              | Base object that has all relevant fields and which uses the Delta mechanism to modify those fields.                                         | Core                                            |

Reviewer 2: ```suggestion
| State              | Object representing data from an experiment, like the conditions, observed experiment data and models. The State also includes rules on how to update those data if new data are provided using the "Delta mechanism".                                           | Core                                            |
```

> | Delta              | An object that determines the extension behavior of State objects.                                                                          | Core                                            |

Reviewer 2: ```suggestion
| Delta              | An object with new data that can be added to a State, returning a new State which includes the new data.                                                                          | Core                                            |
```

| Variables          | A State field that holds experimental variables, which are defined according to name, type, units, allowed values, and range.               | Experimentalists, Experiment Runners, Theorists |
| VariableCollection | Immutable metadata about dependent variables, independent variables, and covariates.                                                        | Experimentalists, Experiment Runners, Theorists |
| Conditions         | A State field that defines what observations should be collected according to a specific combination of values of the independent variables | Experimentalists, Experiment Runners, Theorists |
| Experiment Data    | A State field that holds specified conditions as well as the corresponding observations.                                                    | Experiment Runners, Theorists                   |
| Model              | A State field that holds the the collection of best fit equations produced by theorists.                                                    | Theorists, Experimentalists                     |

> | Components         | The objects that can act on the State (e.g., experimentalists, experiment runners, theorists).                                              | Experimentalists, Experiment Runners, Theorists |

Reviewer 2: Maybe "functions"? Most of what we do is functional now, and that makes things a bit easier for people to understand.
```suggestion
| Components         | The functions that can act on the State (e.g., experimentalists, experiment runners, theorists).                                              | Experimentalists, Experiment Runners, Theorists |
```

> | Components         | The objects that can act on the State (e.g., experimentalists, experiment runners, theorists).                                              | Experimentalists, Experiment Runners, Theorists |

Reviewer 2: Did we settle on calling these "Components"? I don't remember if we said that or "Tasks".

Author: same comment as above

| Experimentalist    | A component that outputs new conditions, which are intended to yield novel observations.                                                    | Experimentalists                                |
| Theorist           | A component that takes in the full collection of conditions and observations and outputs models that link the two.                          | Theorists                                       |
| Experiment Runner  | A component that takes in conditions and collects corresponding observations.                                                               | Experiment Runners                              |
| Wrapper            | Special functions that make the components of AutoRA able to operate on State objects.                                                      | Experimentalists, Experiment Runners, Theorists |
| Workflow           | A collection of tools that enable closed-loop empirical research with the AutoRA framework.                                                 | Experimentalists, Experiment Runners, Theorists |

> | Cycle              | A workflow tool that allows AutoRA components to be chained together in serial loops.                                                       | Experimentalists, Experiment Runners, Theorists |

Reviewer 2: The `Cycle` is completely deprecated at this point – I'd drop it completely.

Author: same comment as above

Author: agree that it makes sense to reflect most recent, but just going off the task as is

> | Cylc               | A workflow engine for cycling systems that orchestrates distributed workflows of interdependent cycling tasks.                              | Experimentalists, Experiment Runners, Theorists |

Reviewer 2: I'd consider dropping "Cylc" here – it's just one way to run a cycle, and I don't think it's the most important way. We touch on Cylc in the specific workflow documentation, and I think that's probably enough.

Author: sounds good to me! per [instruction](https://github.com/AutoResearch/autora/issues/516#issuecomment-1716300791) in the task, i was going off this [page](https://docs.google.com/presentation/d/1OdP4M2ByW8GKTFvcCgDG5HDzUwHKdFEhj1ddH4ClvEw/mobilepresent?slide=id.g27e19158dea_1_0)
